### PR TITLE
webgpu: Ebridgewater/upload texture webgpu

### PR DESCRIPTION
**Pull request by bridgewaterrobbie**
bugs=[410600521]

### CODE REVIEW

File: filament/backend/include/backend/PixelBufferDescriptor.h [hunk 1]
```diff
@@ -251,7 +243,7 @@ class UTILS_PUBLIC PixelBufferDescriptor : public BufferDescriptor {
 
         size_t bpp = n;
```
[the remark this answers is not in the document]
Comment (bridgewaterrobbie) [reply filament/backend/include/backend/PixelBufferDescriptor.h:244]: Yes, that’s my understanding too.

also I can’t reply to your other comment, but I’m testing by stacking this on top of Juan’s changes locally

File: filament/backend/include/backend/PixelBufferDescriptor.h [hunk 2]
```diff
@@ -282,16 +274,35 @@ class UTILS_PUBLIC PixelBufferDescriptor : public BufferDescriptor {
...
+     * @param stride    Stride of a row in pixels
+     * @param height    Height of the image in rows
+     * @param alignment Alignment in bytes of pixel rows
+     * @return The buffer size needed to fit this image in bytes
+     */
+    static constexpr size_t computeDataSize(PixelDataFormat format, PixelDataType type,
+            size_t stride, size_t height, size_t alignment) noexcept {
+        assert_invariant(alignment);
 
+        size_t bpp = computePixelSize(format, type);
```
[the remark this answers is not in the document]
Comment (bridgewaterrobbie) [reply filament/backend/include/backend/PixelBufferDescriptor.h:296]: It’s more that I needed the pixel size and didn’t want to remake it, but that was the only part I needed.

File: filament/backend/include/backend/PixelBufferDescriptor.h [hunk 3]
```diff
@@ -282,16 +274,35 @@ class UTILS_PUBLIC PixelBufferDescriptor : public BufferDescriptor {
                 bpp = 2;
                 break;
         }
+        return bpp;
```
[the remark this answers is not in the document]
Comment (bridgewaterrobbie) [reply filament/backend/include/backend/PixelBufferDescriptor.h:277]: Right, it’s hard to be certain especially without the infra to test and validate. I’m taking an approach of trusting existing common code where it makes sense as well as reading docs, but I think these edge cases will just need to be tested and investigated when we are more easily able to run through them. 

I’m not even fully confident yet in some of my edge case handling here. The most I could do is edit the quad image to be a few pixels short in order to test the padding logic, but it is good enough for now I figure.

File: filament/backend/src/webgpu/WebGPUHandles.cpp [hunk 4]
```diff
@@ -20,7 +20,7 @@
 
 #include <utils/BitmaskEnum.h>
 #include <utils/Panic.h>
-
+#include <private/backend/BackendUtils.h>
```
[the remark this answers is not in the document]
Comment (bridgewaterrobbie) [reply filament/backend/src/webgpu/WebGPUHandles.cpp:23]: Answering from mobile so I can’t check the code , but my thinking was that the WebGPU directory is basically functioning as its own library and thus brackets make sense. Though I think you are right and it’s actually compiled at once, sort of an edge case

File: filament/backend/src/webgpu/WebGPUDriver.cpp [hunk 5]
```diff
@@ -753,11 +753,72 @@ void WebGPUDriver::setVertexBufferObject(Handle<HwVertexBuffer> vbh, uint32_t in
...
+        uint32_t yoffset, uint32_t zoffset, uint32_t width, uint32_t height, uint32_t depth,
+        PixelBufferDescriptor&& p) {
+    PixelBufferDescriptor* data = &p;
+    PixelBufferDescriptor reshapedData;
+    if (reshape(p, reshapedData)) {
+        data = &reshapedData;
+    }
+    auto texture = handleCast<WGPUTexture>(th);
+
+    // TODO: Writing to a depth texture is illegal and errors. I'm not sure why Filament is trying
```
[the remark this answers is not in the document]
Comment (bridgewaterrobbie) [reply filament/backend/src/webgpu/WebGPUDriver.cpp:766]: I think you have what I'm solving backwards- without this check, I was getting errors from Dawn that we were attempting to write to a depth texture, which is illegal. By checking the aspect that we are setting and skipping the upload iff it is a depth texture, the warnings were silenced.

**Comment by bridgewaterrobbie** [comment]
Please hold until both @poweifeng approves, and the dummy-removal PR is merged (I think the primary thing needed there is the samplerDesc fix)

File: filament/backend/src/webgpu/WebGPUDriver.cpp [hunk 6]
```diff
@@ -752,11 +753,72 @@ void WebGPUDriver::setVertexBufferObject(Handle<HwVertexBuffer> vbh, uint32_t in
...
+        .aspect = texture->getAspect() };
+    uint32_t bytesPerRow = static_cast<uint32_t>(
+            PixelBufferDescriptor::computePixelSize(data->format, data->type) * width);
+    auto extent = wgpu::Extent3D{ .width = width, .height = height, .depthOrArrayLayers = depth };
+
+    const uint8_t* dataBuff = static_cast<const uint8_t*>(data->buffer);
+    size_t dataSize = data->size;
+    std::unique_ptr<uint8_t[]> paddedBuffer;
+
+    if (bytesPerRow % 256 != 0) {
```
Comment (jc3265) [inline filament/backend/src/webgpu/WebGPUDriver.cpp:798]: i really doubt that the `256` will change BUT in theory we could define it somewhere as a constant and we would avoid a _magic_ number

File: filament/backend/src/webgpu/WebGPUDriver.cpp [hunk 7]
```diff
@@ -752,11 +753,72 @@ void WebGPUDriver::setVertexBufferObject(Handle<HwVertexBuffer> vbh, uint32_t in
...
+    PixelBufferDescriptor* data = &p;
+    PixelBufferDescriptor reshapedData;
+    if (reshape(p, reshapedData)) {
+        data = &reshapedData;
+    }
+    auto texture = handleCast<WGPUTexture>(th);
+
+    // TODO: Writing to a depth texture is illegal and errors. I'm not sure why Filament is trying
+    // to do so, but early returning is working?
+    if(texture->getAspect() == wgpu::TextureAspect::DepthOnly){
```
Comment (poweifeng) [inline filament/backend/src/webgpu/WebGPUDriver.cpp:768]: Maybe file a bug for this

File: filament/backend/src/webgpu/WebGPUDriver.cpp [hunk 8]
```diff
@@ -752,11 +753,72 @@ void WebGPUDriver::setVertexBufferObject(Handle<HwVertexBuffer> vbh, uint32_t in
...
+        paddedBuffer = std::make_unique<uint8_t[]>(paddedBufferSize);
+        uint8_t* dest = paddedBuffer.get();
+
+        for (uint32_t z = 0; z < depth; ++z) {
+            for (uint32_t y = 0; y < height; ++y) {
+                std::memcpy(dest, dataBuff, bytesPerRow);
+                dest += paddedBytesPerRow;
+                dataBuff += bytesPerRow;
+            }
+        }
```
Comment (poweifeng) [inline filament/backend/src/webgpu/WebGPUDriver.cpp:812]: couldn't you do just one memcpy if padding was 0?

**Review by poweifeng — APPROVED** [review]
PixelBufferDescriptor lgtm.

File: filament/backend/src/webgpu/WebGPUDriver.cpp (hunk 7 above)
Comment (bridgewaterrobbie) [reply filament/backend/src/webgpu/WebGPUDriver.cpp:768]: Sure. For the full project?

Comment (poweifeng) [reply filament/backend/src/webgpu/WebGPUDriver.cpp:768]: not sure what "full project" means, but I mean maybe you can file a bug to indicate that somehow we're trying to write into a depth texture, and file it where you usually do.

File: filament/backend/src/webgpu/WebGPUDriver.cpp (hunk 8 above)
Comment (bridgewaterrobbie) [reply filament/backend/src/webgpu/WebGPUDriver.cpp:812]: Padding can't be 0 if it enters the `if (bytesPerRow % 256 != 0)` block

Comment (poweifeng) [reply filament/backend/src/webgpu/WebGPUDriver.cpp:812]: ah yes, sorry I missed the indent.
